### ISSUE TITLE
Different image components for plugin source and output

### ISSUE THREAD

**Issue by nitsch** [MrKepzie/Natron#648]
Defining a general-context OFX plugin that accepts Alpha, RGB, and RGBA as its source clip but only Alpha as its output clip, Natron will always convert the source clip to an Alpha image, i.e. the plugin never sees an RGB or RGBA source image. Is this intended behaviour?

Tested in the x64 Windows release of Natron 1.2.1.

**Comment by MrKepzie** [comment]
Yes this is from the HostSupport wrapper. This is kind of described vaguely here: http://openfx.sourceforge.net/Documentation/1.3/ofxProgrammingReference.html#id482755

Basically on the plug-in side this is very often expected that the components in output are the same components in input.

**Comment by nitsch** [comment]
> http://openfx.sourceforge.net/Documentation/1.3/ofxProgrammingReference.html#id482755

Hm, this seems to be talking mostly about harmonizing different
*inputs*, not about making input and output the same component-type.


> Basically on the plug-in side this is very often expected that the
> components in output are the same components in input.

Are you aware of any other place in the specs that touches this topic?
It seems rather strong an expectation without the specs explicitly
saying so.

Nuke for instance does not seem to care much about this and will happily
feed the described plugin RGB/RGBA and request the offered Alpha as output.

**Comment by MrKepzie** [comment]
The documentation is a little sparse on this, but basically this is done in Natron in OfxEffectInstance.cpp:

       ///Set the clip to have the same components as the output components if it is supported
        if ( clip->isSupportedComponent(foundOutputPrefs->second.components) ) {
            ///we only take into account non mask clips for the most components
            if ( !clip->isMask() && (foundClipPrefs->second.components != foundOutputPrefs->second.components) ) {
                foundClipPrefs->second.components = foundOutputPrefs->second.components;
                hasChanged = true;
            }
        }

This was done so because most of the plug-ins existing expect the same components in Input and Output. Unfortunately there is no way in OpenFX to tell the host that you do not support multiple clips components and assuming the plug-in does not support it was the safest bet so far. We could try to come up with a smarter solution.

Basically for now all our plug-ins in openfx-misc do the following check:

    // see if they have the same depths and bytes and all
        if (srcBitDepth != dstBitDepth || srcComponents != dstComponents) {
            OFX::throwSuiteStatusException(kOfxStatErrImageFormat);
        }

**Comment by devernay** [comment]
For the record, AFAIK Nuke doesn't support RGB.
Only Alpha and RGBA.

**Comment by nitsch** [comment]
> For the record, AFAIK Nuke doesn't support RGB.
> Only Alpha and RGBA.

That's quite possible. It doesn't properly work with eBitDepthUByte either
(at least Nuke 9). Claims to pass the plugin that type but the data is
actually a eBitDepthFloat buffer.

**Comment by nitsch** [comment]
> This was done so because most of the plug-ins existing expect the same
> components in Input and Output.

Do you mean available OFX plugins in general or the Natron-related ones?
Because I would assume that most of the commercial ones have been tested
with Nuke (if not even primarily developed for this host) which, like I
said, seems to be happily feeding RGBA and requesting Alpha as output.


> Unfortunately there is no way in OpenFX to
> tell the host that you do not support multiple clips components and
> assuming the plug-in does not support it was the safest bet so far. We
> could try to come up with a smarter solution.

Maybe this is something that should not be handled at the plugin level but
in the host instead. Maybe an additional option in Natron: "Force same
input and output clip components if possible". I am not sure.

**Comment by MrKepzie** [comment]
Alexander, I'm going to deactivate the lines I pasted to you in the code above and experiment to see if that breaks something. I added these lines back in the days where our plug-ins did not implement getClipPreference action correctly

**Comment by devernay** [comment]
The default behavior in [Instance::setDefaultClipPreferences()](https://github.com/devernay/openfx/blob/master/HostSupport/src/ofxhImageEffect.cpp#L1994) seems reasonable, and Natron should not force the input components given the output components.
- The default value for outputcomponents is computed from the most chromatic input components, intersected with the plugin capabilities
- If not happy with that default value, the plugin may set the output components on its output, given the input components, in [kOfxImageEffectActionGetClipPreferences](http://openfx.sourceforge.net/Documentation/1.3/ofxProgrammingReference.html#ImageEffectClipPreferences)

This issue is being worked on.

**Comment by MrKepzie** [comment]
This is pushed, you should try your plug-in on the workshop git head and tell us if that changes something for your plug-in

**Comment by nitsch** [comment]
> This is pushed, you should try your plug-in on the workshop git head and
> tell us if that changes something for your plug-in

I do not have a Natron build environment set up, so it will probably take
me a while to do this. But I will let you know when I have tested your
changes.

**Comment by nitsch** [comment]
I can confirm that your changes do have the desired effect. The input is no longer forced to the same components as the output.
